### PR TITLE
Implement ActionInterface for /catalog/category/view

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
This PR is example for **How to replace Abstract Action class with ActionInterface**

The performance change after:
image

### Related Pull Requests
* https://github.com/magento/partners-magento2ee/pull/157

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)
N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magento run all tests

Tests failed due to relation with Magento Commerce. Changes to the plugin in Magento Commerce were introduced and now Tests should pass.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Adjusted code to follow the Coding Standards in Magento EE

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz,
According to test failures - looks like your change broke category page.
Please review them and adjust your changes

**Comment by lbajsarowicz** [comment]
@ihor-sviziev I've introduced `strict_types` to the tests, that caused failures on types juggling.
The easies way is to rollback the `strict_types`. Later on, we need to review these tests and change the call arguments to be correct type.

**Comment by lbajsarowicz** [comment]
@magento run all tests

@ihor-sviziev I've introduced `strict_types` to the tests, that caused failures on types juggling.
The easies way is to rollback the `strict_types`. Later on, we need to review these tests and change the call arguments to be correct type.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
The issue
```
[2020-03-18 14:18:37] report.CRITICAL: Error: Call to undefined method Magento\Catalog\Controller\Category\View\Interceptor::getResponse() in /var/www/html/.ee/app/code/Magento/CatalogPermissions/Observer/ApplyCategoryPermissionObserver.php:109
Stack trace:
#0 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(88): Magento\CatalogPermissions\Observer\ApplyCategoryPermissionObserver->execute(Object(Magento\Framework\Event\Observer))
#1 /var/www/html/lib/internal/Magento/Framework/Event/Invoker/InvokerDefault.php(74): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod(Object(Magento\CatalogPermissions\Observer\ApplyCategoryPermissionObserver), Object(Magento\Framework\Event\Observer))
#2 /var/www/html/.ee/app/code/Magento/Staging/Model/Event/Manager.php(97): Magento\Framework\Event\Invoker\InvokerDefault->dispatch(Array, Object(Magento\Framework\Event\Observer))
#3 /var/www/html/generated/code/Magento/Staging/Model/Event/Manager/Proxy.php(95): Magento\Staging\Model\Event\Manager->dispatch('catalog_control...', Array)
#4 /var/www/html/app/code/Magento/Catalog/Controller/Category/View.php(232): Magento\Staging\Model\Event\Manager\Proxy->dispatch('catalog_control...', Array)
#5 /var/www/html/app/code/Magento/Catalog/Controller/Category/View.php(256): Magento\Catalog\Controller\Category\View->_initCategory()
#6 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\Catalog\Controller\Category\View->execute()
#7 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\Catalog\Controller\Category\View\Interceptor->___callParent('execute', Array)
#8 /var/www/html/lib/internal/Magento/Framework/App/Action/Plugin/ActionFlagNoDispatchPlugin.php(51): Magento\Catalog\Controller\Category\View\Interceptor->Magento\Framework\Interception\{closure}()
#9 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\Framework\App\Action\Plugin\ActionFlagNoDispatchPlugin->aroundExecute(Object(Magento\Catalog\Controller\Category\View\Interceptor), Object(Closure))
#10 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Catalog\Controller\Category\View\Interceptor->Magento\Framework\Interception\{closure}()
#11 /var/www/html/generated/code/Magento/Catalog/Controller/Category/View/Interceptor.php(26): Magento\Catalog\Controller\Category\View\Interceptor->___callPlugins('execute', Array, Array)
#12 /var/www/html/lib/internal/Magento/Framework/App/FrontController.php(164): Magento\Catalog\Controller\Category\View\Interceptor->execute()
#13 /var/www/html/lib/internal/Magento/Framework/App/FrontController.php(98): Magento\Framework\App\FrontController->processRequest(Object(Magento\Framework\App\Request\Http), Object(Magento\Catalog\Controller\Category\View\Interceptor))
#14 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#15 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#16 /var/www/html/app/code/Magento/Store/App/FrontController/Plugin/RequestPreprocessor.php(99): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#17 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#18 /var/www/html/app/code/Magento/PageCache/Model/App/FrontController/BuiltinPlugin.php(73): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#19 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#20 /var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#21 /var/www/html/generated/code/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, NULL)
#22 /var/www/html/lib/internal/Magento/Framework/App/Http.php(116): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#23 /var/www/html/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#24 /var/www/html/lib/internal/Magento/Framework/App/Bootstrap.php(260): Magento\Framework\App\Http\Interceptor->launch()
#25 /var/www/html/pub/index.php(40): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#26 {main} [] []
```
is caused by B2B module. Addressing the issue

**Comment by lbajsarowicz** [comment]
The problem with B2B is much much more complicated.
It's related with "Magic" observer that changes:
1. Category model after loading
2. Replaces the response

I need to redesign whole Observer (`app/code/Magento/CatalogPermissions/Observer/ApplyCategoryPermissionObserver.php`) into 2 plugins.


The funny fact is that this PR was to show how easy is to migrate from `AbstractAction` to `ActionInterface`. I forgot that there are people who do not follow the [most basic Magento rule](https://devdocs.magento.com/guides/v2.3/ext-best-practices/extension-coding/observers-bp.html) when developing Magento B2B.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

Removed Unit Tests for Observer in EE module.

**Comment by lbajsarowicz** [comment]
@magento run all tests
1. Added Unit Tests
2. Fixed Static Tests

**Comment by ihor-sviziev** [comment]
@lbajsarowicz tests still failing, and it's not only one test, so seems like changes breaks functionality

**Comment by lbajsarowicz** [comment]
I've done big changes to Magento commerce code. Still needs to be adjusted.

**Comment by lbajsarowicz** [comment]
@magento run all tests

Fixed some more static tests. I'm not sure about the failure that is happening with the redirect - as Xdebug shows me that the default values for categories were changed somehow during the test (looks like 2 tests affecting each other). I'm going to investigate that during the night.

**Comment by lbajsarowicz** [comment]
@magento run all tests

We need to redesign View controller :-1:

**Comment by lbajsarowicz** [comment]
@magento run all tests

Changes to EE

**Comment by lbajsarowicz** [comment]
@magento run all tests
CI took the old version of Commerce code.

**Comment by lbajsarowicz** [comment]
@lenaorobei / @slavvka 

There's something wrong with CI:
image

The code:
image

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7168](https://jira.corp.magento.com/browse/ENGCOM-7168) has been created to process this Pull Request

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Test/Unit/Controller/Category/ViewTest.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Controller/Category/ViewTest.php]: Could you add return type?
```suggestion
    protected function setUp(): void
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@lbajsarowicz could you fix static test failure?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7168](https://jira.corp.magento.com/browse/ENGCOM-7168) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
Moved to "on hold" for the same reason as described here: https://github.com/magento/magento2/pull/27538#issuecomment-624891873

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Test/Unit/Controller/Category/ViewTest.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Controller/Category/ViewTest.php:6]: Looks like we should bring it back

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz ,
We finally have final decision, it clearly described here https://github.com/magento/magento2/pull/27296#pullrequestreview-418541438.

Could you update your PR to make it mergeble again and revert not needed changes, for instance removing strict types?

**Comment by lbajsarowicz** [comment]
If anyone else wants to pick this, feel free to take my code and continue.
I won't be available to handle that in next 2 - 3 weeks.

**Pull request closed without merging**
